File: bindings/dotnet/src/Pamoja.Profile/Rules.cs
using System.Text.Json;

using Pamoja.Kit;
using Pamoja.Native.Interop;

namespace Pamoja.Profile;

/// <summary>What a rule calls for when its condition sets or clears.</summary>
public enum RuleActionKind
{
    /// <summary>Switch the output the program holds under <see cref="RuleAction.Actuator"/>.</summary>
    Drive = 1,

    /// <summary>Publish <see cref="RuleAction.Payload"/> to <see cref="RuleAction.Topic"/>.</summary>
    Publish = 2,
}

/// <summary>One thing a rule calls for, as the rule file writes it.</summary>
/// <remarks>Only the values belonging to <see cref="Kind"/> are set.</remarks>
/// <param name="Kind">Whether to switch an output or publish a message.</param>
/// <param name="Actuator">The output to switch, for a drive.</param>
/// <param name="On">The setting to switch it to, for a drive.</param>
/// <param name="Topic">The topic to publish to, for a publish.</param>
/// <param name="Payload">The text to publish, for a publish.</param>
public readonly record struct RuleAction(
    RuleActionKind Kind,
    string? Actuator,
    bool? On,
    string? Topic,
    string? Payload);

/// <summary>What one rule did with one reading.</summary>
/// <param name="Rule">The rule's name.</param>
/// <param name="Edge">Whether its condition set or cleared.</param>
/// <param name="Reading">The reading that did it.</param>
/// <param name="Actions">
/// What the rule calls for on this edge: its <c>then</c> actions when it set and its
/// <c>otherwise</c> actions when it cleared, in the order the file gives them.
/// </param>
public sealed record RuleFired(
    string Rule,
    Edge Edge,
    float Reading,
    IReadOnlyList<RuleAction> Actions);

/// <summary>
/// The decisions a rule file makes, reading by reading, with no link and no outputs of
/// its own.
/// </summary>
/// <remarks>
/// The program moves the messages and drives the outputs: it hands each reading over
/// with the topic it arrived on, and learns which rules set or cleared and what each
/// calls for. This is the deciding half of the Rust <c>RuleEngine</c>, which owns a link
/// and outputs and so stays in Rust; the two decide alike.
/// </remarks>
public sealed class RuleEvaluator : IDisposable
{
    private readonly NativeHandle _handle;

    private RuleEvaluator(IntPtr handle) =>
        _handle = NativeHandle.Create(handle, NativeMethods.pamoja_rule_evaluator_free, "rule evaluator");

    /// <summary>Loads a rule file and arms its rules, every condition starting cleared.</summary>
    /// <param name="text">The rule file.</param>
    /// <returns>The evaluator, which the caller disposes.</returns>
    /// <exception cref="PamojaException">
    /// The text is not a rule file, or holds a rule no engine could run, such as one that
    /// watches a filter or has nothing to do.
    /// </exception>
    public static RuleEvaluator FromJson(string text) =>
        new(NativeMethods.pamoja_rule_evaluator_from_json(text));

    /// <summary>Gets the topics the rules watch, each once, in name order: the topics to subscribe to.</summary>
    public IReadOnlyList<string> Topics =>
        _handle.Use(rules => Names(NativeMethods.pamoja_rule_evaluator_topics_json(rules)));

    /// <summary>Gets the actuators the rules drive, each once, in name order: the outputs the program has to have.</summary>
    public IReadOnlyList<string> Actuators =>
        _handle.Use(rules => Names(NativeMethods.pamoja_rule_evaluator_actuators_json(rules)));

    /// <summary>Writes the rules back out as the file a fleet shares.</summary>
    /// <returns>The rule file.</returns>
    public string ToJson() =>
        _handle.Use(rules => OwnedString.Read(NativeMethods.pamoja_rule_evaluator_to_json(rules)));

    /// <summary>Judges one reading from one topic against every rule that watches it.</summary>
    /// <param name="topic">The topic the reading arrived on.</param>
    /// <param name="reading">The reading, already decoded from the message.</param>
    /// <returns>
    /// What fired, in the order the rules are listed; empty when no rule watches the topic
    /// or the reading changed nothing.
    /// </returns>
    /// <exception cref="PamojaException">
    /// A rule watches the topic and the reading is not a finite number.
    /// </exception>
    public IReadOnlyList<RuleFired> Evaluate(string topic, float reading) =>
        _handle.Use(rules =>
        {
            string json = OwnedString.Read(
                NativeMethods.pamoja_rule_evaluator_evaluate(rules, topic, reading));
            using JsonDocument document = JsonDocument.Parse(json);
            var fired = new List<RuleFired>();
            foreach (JsonElement one in document.RootElement.EnumerateArray())
            {
                var actions = new List<RuleAction>();
                foreach (JsonElement action in one.GetProperty("actions").EnumerateArray())
                {
                    actions.Add(ActionOf(action));
                }

                fired.Add(new RuleFired(
                    one.GetProperty("rule").GetString()!,
                    one.GetProperty("edge").GetString() == "set" ? Edge.Set : Edge.Cleared,
                    one.GetProperty("reading").GetSingle(),
                    actions));
            }

            return fired;
        });

    /// <summary>Reports whether any rule watches a topic, so the program knows whether to decode a message.</summary>
    /// <param name="topic">The topic a message arrived on.</param>
    /// <returns><c>true</c> when some rule watches it.</returns>
    public bool Watches(string topic) =>
        _handle.Use(rules => NativeMethods.pamoja_rule_evaluator_watches(rules, topic));

    /// <summary>Reports whether a rule's condition currently holds.</summary>
    /// <param name="rule">The rule's name.</param>
    /// <returns>The condition's state, or <c>null</c> for a name no rule has.</returns>
    public bool? IsSet(string rule) =>
        _handle.Use(rules =>
            NativeMethods.pamoja_rule_evaluator_is_set(rules, rule, out bool set) == PamojaStatus.Ok
                ? set
                : (bool?)null);

    /// <inheritdoc/>
    public void Dispose() => _handle.Dispose();

    /// <summary>Reads one action as the rule file writes it.</summary>
    private static RuleAction ActionOf(JsonElement action) =>
        action.GetProperty("do").GetString() == "drive"
            ? new RuleAction(
                RuleActionKind.Drive,
                action.GetProperty("actuator").GetString(),
                action.GetProperty("on").GetBoolean(),
                null,
                null)
            : new RuleAction(
                RuleActionKind.Publish,
                null,
                null,
                action.GetProperty("topic").GetString(),
                action.GetProperty("payload").GetString());

    /// <summary>Reads a JSON array of names the native side handed back.</summary>
    private static IReadOnlyList<string> Names(IntPtr json)
    {
        using JsonDocument document = JsonDocument.Parse(OwnedString.Read(json));
        return document.RootElement.EnumerateArray().Select(name => name.GetString()!).ToList();
    }
}

File: examples/guides/rules.rs
//! The rules guide example; see docs/guides/rules.md.
//!
//! Run: `cargo run -p pamoja-examples --example rules`

use std::error::Error;

use std::sync::{Arc, Mutex};

use pamoja_core::{Actuator, Result};

/// The valve on the other node, recording every switch so the example can count them.
#[derive(Clone, Default)]
struct Valve {
    switches: Arc<Mutex<Vec<bool>>>,
}

impl Actuator for Valve {
    type Command = bool;

    async fn apply(&mut self, open: bool) -> Result<()> {
        self.switches.lock().expect("valve lock").push(open);
        Ok(())
    }
}

/// A rule file that reads one node's topic and drives another node's valve, run by the
/// engine off an in-process broker; then the same file judged by an evaluator, and what
/// goes wrong.
#[tokio::main]
async fn main() -> std::result::Result<(), Box<dyn Error>> {
    // ANCHOR: example
    use pamoja_codec::JsonCodec;
    use pamoja_core::{Receive, Transport};
    use pamoja_kit::Edge;
    use pamoja_loopback::{LoopbackBroker, LoopbackTransport};
    use pamoja_profile::{Action, RuleEngine, Rules};

    // A rule is a file: the topic it watches, the line a reading crosses, the release
    // band that stops it firing over and over, and what to do on the way down and on the
    // way back. Two rules watch one bed here: one waters it when it dries past 30 and
    // stops once it is wetter than 35, and one raises an alarm when it is soaked past 60.
    let file = r#"{ "rules": [
  { "name": "water-when-dry",
    "when": { "topic": "garden/bed-1/moisture", "below": 30.0, "hysteresis": 5.0 },
    "then": [ { "drive": "bed-valve", "on": true },
              { "publish": "garden/bed-1/valve", "payload": "open" } ],
    "otherwise": [ { "drive": "bed-valve", "on": false },
                   { "publish": "garden/bed-1/valve", "payload": "closed" } ] },
  { "name": "flood-alarm",
    "when": { "topic": "garden/bed-1/moisture", "above": 60.0, "hysteresis": 5.0 },
    "then": [ { "publish": "garden/alarm", "payload": "waterlogged" } ] }
] }"#;

    // Three parties on one broker: the node that reads the bed, the engine that holds the
    // valve, and a watcher on the topics the rules publish to.
    let broker = LoopbackBroker::new();
    let mut probe = LoopbackTransport::new(broker.clone());
    let mut watcher = LoopbackTransport::new(broker.clone());
    probe.connect().await?;
    watcher.connect().await?;
    watcher.subscribe("garden/bed-1/valve").await?;
    watcher.subscribe("garden/alarm").await?;

    let valve = Valve::default();
    let mut engine = RuleEngine::new(
        Rules::from_json(file)?,
        LoopbackTransport::new(broker),
        JsonCodec,
    )
    .with_actuator("bed-valve", valve.clone());
    engine.connect().await?;
    println!(
        "watches   {}, and drives {}",
        engine.topics().join(", "),
        engine.actuators().join(", ")
    );

    // The bed dries out, is watered, and floods. A rule fires only as its condition sets
    // or clears, and the readings in between change nothing. At 65 two rules fire on one
    // reading, in the order the file lists them.
    for reading in [42.0f32, 31.0, 28.0, 33.0, 65.0, 50.0] {
        probe
            .send_text("garden/bed-1/moisture", &reading.to_string())
            .await?;
        let fired = engine.step().await?.expect("the link is up");
        if fired.is_empty() {
            println!("{reading:<10}nothing fired");
        }
        for one in &fired {
            let edge = match one.edge {
                Edge::Set => "set",
                Edge::Cleared => "cleared",
            };
            let actions: Vec<String> = one
                .actions
                .iter()
                .map(|action| match action {
                    Action::Drive { actuator, on } => {
                        format!("drive {actuator} {}", if *on { "on" } else { "off" })
                    }
                    Action::Publish { topic, payload } => format!("publish {payload} to {topic}"),
                })
                .collect();
            if actions.is_empty() {
                println!("{reading:<10}{} {edge}, with nothing to do", one.rule);
            } else {
                println!("{reading:<10}{} {edge}: {}", one.rule, actions.join(", "));
            }
        }
    }

    // The watcher heard every message the rules published, in the order they went out.
    let mut heard = Vec::new();
    for _ in 0..3 {
        let message = watcher.recv().await?.expect("a message");
        heard.push(message.text().expect("words").to_owned());
    }
    println!("heard     {}", heard.join(", "));
    let switches = valve.switches.lock().expect("valve lock").clone();
    let state = if switches.last() == Some(&true) {
        "on"
    } else {
        "off"
    };
    println!(
        "valve     switched {} times, and it is {state}",
        switches.len()
    );
    // ANCHOR_END: example

    assert_eq!(heard, ["open", "closed", "waterlogged"]);
    assert_eq!(switches, [true, false]);
    assert_eq!(engine.is_set("water-when-dry"), Some(false));
    assert_eq!(engine.is_set("flood-alarm"), Some(false));

    // ANCHOR: wrong
    use pamoja_profile::RuleEvaluator;

    // A program that moves its own messages hands each reading to an evaluator, the
    // engine's deciding half on its own, and carries out what it says.
    let mut evaluator = RuleEvaluator::new(Rules::from_json(file)?)?;

    // A reading that is not a number, such as the NaN a failed probe reports, is refused
    // on a watched topic rather than leaving every rule as it was with nothing to say why.
    match evaluator.evaluate("garden/bed-1/moisture", f32::NAN) {
        Ok(_) => println!("a reading of NaN was judged, which should never happen"),
        Err(error) => println!("refused   {error}"),
    }

    // A topic no rule watches is not judged at all, so even a NaN there says nothing.
    if evaluator
        .evaluate("garden/bed-2/moisture", f32::NAN)?
        .is_empty()
    {
        println!(
            "ignored   no rule watches garden/bed-2/moisture, so even a NaN there is not judged"
        );
    }

    // A file no engine could run is refused as it loads, with the rule and the reason.
    for edited in [
        file.replace("garden/bed-1/moisture", "garden/+/moisture"),
        file.replace("\"flood-alarm\"", "\"water-when-dry\""),
    ] {
        match Rules::from_json(&edited).and_then(RuleEvaluator::new) {
            Ok(_) => println!("a file no engine could run was accepted, which should never happen"),
            Err(error) => println!("refused   {error}"),
        }
    }

    // With no release band, readings that hover at the line set and clear the rule on
    // every sample, and each edge switches the valve. The band of 5 holds it through them.
    let fires = |text: &str| -> std::result::Result<usize, Box<dyn Error>> {
        let mut judge = RuleEvaluator::new(Rules::from_json(text)?)?;
        let mut count = 0;
        for reading in [29.9, 30.1, 29.8, 30.2] {
            count += judge.evaluate("garden/bed-1/moisture", reading)?.len();
        }
        Ok(count)
    };
    let bare = fires(&file.replace("\"hysteresis\": 5.0", "\"hysteresis\": 0.0"))?;
    let banded = fires(file)?;
    println!(
        "chatter   4 readings hovering at 30 fire the rule {bare} times with no release band, {banded} with a band of 5"
    );
    // ANCHOR_END: wrong

    Ok(())
}
